sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -873,24 +873,30 @@ object InferFiltersFromGenerate extends Rule[LogicalPlan] {
       if !e.deterministic || e.children.forall(_.foldable) => generate
 
     case generate @ Generate(g, _, false, _, _, _) if canInferFilters(g) =>
-      // Exclude child's constraints to guarantee idempotency
-      val inferredFilters = ExpressionSet(
-        Seq(
-          GreaterThan(Size(g.children.head), Literal(0)),
-          IsNotNull(g.children.head)
-        )
-      ) -- generate.child.constraints
-
-      if (inferredFilters.nonEmpty) {
-        generate.copy(child = Filter(inferredFilters.reduce(And), generate.child))
-      } else {
-        generate
+      g.children.head match {
+        case _: CreateNonEmptyNonNullCollection =>
+          // we don't need to add filters when creating an array because we know its size
+          // is > 0 and its not null
+          generate
+        case _ =>
+          // Exclude child's constraints to guarantee idempotency
+          val inferredFilters = ExpressionSet(
+            Seq(
+              GreaterThan(Size(g.children.head), Literal(0)),
+              IsNotNull(g.children.head)
```

Comment:
In general, optimizer rules should be orthogonal. For this case, I  think a better idea is to add a new optimizer rule, which optimizes `IsNotNull` and size check expressions above `CreateArray/Map` into true literal.

Reply:
@cloud-fan, sorry, I'm not following what exactly you are suggesting, could you clarify?  Are you saying creating a new Rule outside of InferFiltersFromGenerate? Are you somehow saying just change it so that inferredFilters come out to be empty?

Reply:
I mean, this rule can add redundant predicates, and we just need another rule to optimize them out. This is how catalyst rules should interact: be orthogonal and focus on one thing.

Actually, `CreateArray.nullable` is false, so `IsNotNull(CreateArray(...))` will be optimized to `true` already, in rule `NullPropagation`. We can probably update `SimplifyBinaryComparison` and add
```
case GreaterThan(Size(_: CreateNonNullCollection), IntegerLiteral(0)) => TrueLiteral
```

Reply:
I mean, this rule can add redundant predicates, and we just need another rule to optimize them out. This is how catalyst rules should interact: be orthogonal and focus on one thing.

Actually, `CreateArray.nullable` is false, so `IsNotNull(CreateArray(...))` will be optimized to `true` already, in rule `NullPropagation`. We can probably update `SimplifyBinaryComparison` and add
```
case GreaterThan(Size(_: CreateNonNullCollection), IntegerLiteral(0)) => TrueLiteral
```

Reply:
thanks for explaining. I'm new to optimizer code. The problem here is that the size is only 0 in this context because otherwise it would be foldable and hit the case statement above.  Since the SimplifyBinaryCompare check is totally separate, is it safe to assume anything with type CreateNonNullCollection then size == 0 can be removed? It definitely applies here, but for instance if someone else adds a separate optimizer rule in the future.  If you think that is ok I can add it there instead?  Let me know what you think

Reply:
The semantic of an expression is decided by the expression itself, not by where we create the expression.

Looking at it again, I think what's 100% safe is we can turn `Size(CreateArray)` to `CreateArray.children.length`, e.g.
```
case Size(c: CreateArray) => Literal(c.children.length)
case Size(c: CreateMap) => Literal(c.children.length / 2)
```
We can put this in rule `ConstantFolding`.

Reply:
that makes sense, I'll update the pr, thanks